pydicom/fileset.py

```diff
@@ -0,0 +1,2742 @@
+            as :class:`str` or :class:`pathlib.Path`.
+        include_orphans : bool, optional
+            If ``True`` (default) include instances referenced by orphaned
+            directory records in the File-set.
+        raise_orphans : bool, optional
+            If ``True`` then raise an exception if orphaned directory records
+            are found in the File-set (default ``False``).
+        """
+        if isinstance(ds_or_path, Dataset):
+            ds = ds_or_path
+        else:
+            ds = dcmread(ds_or_path)
+
+        sop_class = ds.file_meta.get("MediaStorageSOPClassUID", None)
+        if sop_class != MediaStorageDirectoryStorage:
+            raise ValueError(
+                "Unable to load the File-set as the supplied dataset is "
+                "not a 'Media Storage Directory' instance"
+            )
+
+        tsyntax = ds.file_meta.TransferSyntaxUID
+        if tsyntax != ExplicitVRLittleEndian:
+            warnings.warn(
+                "The DICOMDIR dataset uses an invalid transfer syntax "
+                f"'{tsyntax.name}' and will be updated to use 'Explicit VR "
+                "Little Endian'"
+            )
+
+        try:
+            path = Path(ds.filename).resolve(strict=True)
+        except FileNotFoundError:
+            raise FileNotFoundError(
+                "Unable to load the File-set as the 'filename' attribute "
+                "for the DICOMDIR dataset is not a valid path: "
+                f"{ds.filename}"
+            )
+        except TypeError:
+            # Custom message if DICOMDIR from bytes, etc
+            raise TypeError(
+                "Unable to load the File-set as the DICOMDIR dataset must "
+                "have a 'filename' attribute set to the path of the "
+                "DICOMDIR file"
+            )
+
+        self.clear()
+        self._id = ds.FileSetID
+        self._uid = ds.file_meta.MediaStorageSOPInstanceUID
+        self._descriptor = ds.get("FileSetDescriptorFileID", None)
+        self._charset = ds.get(
+            "SpecificCharacterSetOfFileSetDescriptorFile", None
+        )
+        self._path = path.parent
+        self._ds = ds
+
+        # Create the record tree
+        self._parse_records(ds, include_orphans, raise_orphans)
+
+        bad_instances = []
+        for instance in self:
+            # Check that the referenced file exists
+            file_id = instance.node._file_id
+            if file_id is None:
+                bad_instances.append(instance)
+                continue
+
+            (self.path / file_id).resolve(strict=True)
+            # If the existing directory structure doesn't match pydicom
+            #   then stage everything to be moved to the new structure
+            file_id = instance.FileID.split(os.path.sep)
+            if instance.ReferencedFileID != file_id:
+                instance._apply_stage('~')
+                self._stage['~'][instance.SOPInstanceUID] = instance
```

Comment:
I think the FileSet movement stage can be simplified now to True/False. Just have the FileInstances check if their File ID == Referenced File ID